packages/babel-parser/test/helpers/to-contextual-syntax-error.js

```diff
@@ -0,0 +1,40 @@
+import { codeFrameColumns } from "@babel/code-frame";
+import { runInThisContext } from "vm";
+
+export default function toContextualSyntaxError(
```

Comment:
Q: What do we need this for?

Reply:
So, what we actually wish we had was [FireFox's `Error.fileName`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/fileName), but unfortunately v8 doesn't support this property. So we use `vm.runInContext` to make it think it's coming from this file. This would on its own though make it appear at the bottom of the trace, not the top, because we're really  doing something weird here combining two ideas of stack trace. So instead of fiddling with that, we have the "Contextual Error", which pretends to exist in "filename", and that has as its cause the Babel Error, that way we get the "at filename" followed by the Babel stacktrace for free from node's inspect, vs. having to do string manipulation (which is more fragile and starts to break with more complicated messages).

Reply:
Rather than string manipulation we could have used `Error.captureStackTrace`, but this approach is good too. However, please add a comment explaining this directly in the source code :pray:

Reply:
Unfortunately `captureStackTrace` doesn't work for this (unless I'm missing part of the API). The goal is to convince the constructor that it's happening inside of a file that it's not actually happening in (`babel-parser/test/.../input.js`), and all I think `captureStackTrace` can do is remove the last function on the stack from showing up (a different problem). Hence the necessary roundtrip through vm to convince node we're "in" that, kind of how node does it themselves: https://github.com/nodejs/node/blob/e46c680bf2b211bbd52cf959ca17ee98c7f657f5/test/message/vm_display_syntax_error.js#L29

Reply:
And yeah, I'll put a comment in there.

Reply:
Done: https://github.com/babel/babel/pull/14201/commits/f6aeb0651250a51ec6b0838a6525a8364e96b82c